### PR TITLE
Handle plugin generated content and config

### LINKED ISSUE #78: Allow `mkdocs-gen-files` generated content
:wave: Loving the plugin! 

Currently also loving https://github.com/oprypin/mkdocs-gen-files, but noticed that if I generate navs, or pages referenced by my static navs, they don't end up making it to the generated docsite.

E.g.
`mkdocs.yml`
```yaml
plugins:
  - gen-files:
      scripts:
        - generate_reference_docs.py
  - awesome-pages:
      filename: .nav.yaml
```

`generate_reference_docs.py`
```python
with mkdocs_gen_files.open(f"reference/global-options.md", "w") as f:
    f.write("Hello world!\n")
with mkdocs_gen_files.open(f"reference/goals/index.md", "w") as f:
    f.write("# Goals!\n")
with mkdocs_gen_files.open(f"reference/goals/.nav.yaml", "w") as f:
    f.write("nav:\n  - index.md\n")
```

along with:

```console
$ tree -a docs/markdown/reference
docs/markdown/reference
└── .nav.yaml
$ cat docs/markdown/reference/.nav.yaml 
nav:
  - Global Options: global-options.md
  - Goals: goals
```

Results in `Reference` being in the nav, but is "flat". No expand arrow.

Additionally, the pages _do_ exist, `/reference/global-options` is a valid URL and gives what I expect.

### PR DESCRIPTION
**Pull request by thejcannon**
This change allows `mkdocs-awesome-pages-plugin` to play nicely with `mkdocs-gen-files`, specifically with generated content (`.md`) and generated config (`.pages`).

This is accomplished by:

1. Defining an `on_files` in the plugin that synthesizes a `File` per config in the docs dir. This means we can later unconditionally pull from `Files` for both in-docs and generated configs.
2. Passing the `Files` down the chain to `NavigationMeta`
	- We can use the `Files` to validate the "out-of-docs" check, so we also remove the `docs_dir` arg
	- Additionally, generated files _aren't_ "in-docs", so that check would mean we did't add the generated docs. So removing it is vital to the overall change anyways
3. Making `_gather_metadata` operate on a "relative file" abstraction (e.g. the path below "docs_dir)
	- For `Page`s, that means adding `item.file.src_path` instead of `item.file.abs_src_path`
	- For sections, the metadata loading gets tricky. We have to load the _absolute_ path (because that's how the generated files work), but they path should be the _relative_ path so we match correctly later.
4. Replace `Meta.try_load_from` with `Meta.try_load_from_files` such that we always load from `Files`
	- But as mentioned, if we do successfully load, we need to swap the `path` attribute with the _relative path_

That summarizes the package changes, then there was a few test changes top facilitate this change:

- All the necessary refactoring to handle the swap from `docs_dir` to `files`
- `test_meta.py` changes for the correct `Files` lookup, and new method
- New test infrastructure to have a test define additional Mkdocs plugins, specifically so we can simulate `mkdocs-gen-files`
- New tests for the specific use-case of generated files

Fixes #78

### CODE REVIEW

**Comment by thejcannon** [comment]
@lukasgeiter This is ready to review.

The je ne sais quoi is boiled down to:
- Try to refer to everything by a "relative path"
- Synthesizing a `File` per config, so there isn't a split codepath

**Comment by thejcannon** [comment]
And to report, this is working as expected on my project :tada:

**Comment by lukasgeiter** [comment]
Thanks for creating this PR! Before I dive into reviewing, I've noticed some formatting changes that shouldn't be there. Please make sure to format the files using [black](https://black.readthedocs.io/en/stable/index.html).

**Comment by lukasgeiter** [comment]
Please disregard the failing tests for pypy, they have nothing to do with your PR. I've actually just removed them on `master` so a rebase should get rid of them. The same does not apply for the failing tests on windows, that seems to be an actual issue 😉

**Comment by thejcannon** [comment]
Thanks for taking a look!

Yeah looks like editor ran black with a different line length (shorter) leading to some spurious reformats. I'll fix those

**Comment by thejcannon** [comment]
@lukasgeiter All tests are passing! Ready for review when you are.

File: mkdocs_awesome_pages_plugin/tests/navigation/test_meta.py [hunk 1]
```diff
@@ -45,34 +50,68 @@ def setUp(self):
...
+                            os.path.dirname(item.file.abs_src_path),
+                        )
+                    )
+                    files.append(item.file)
+                elif isinstance(item, Section):
+                    populate_config_dirs(item.children)
+
+        populate_config_dirs(items)
+
+        for config_dir, docs_dir in config_dirs:
```
Comment (lukasgeiter) [inline mkdocs_awesome_pages_plugin/tests/navigation/test_meta.py:77]: Something with `docs_dir` is off. Either the name is misleading or something on line 68 is wrong. I would argue the dirname of the absolute source path isn't always the same as the src_dir.

**Review by lukasgeiter — CHANGES_REQUESTED** [review]
Great work! I especially appreciate the added test cases ❤️

File: mkdocs_awesome_pages_plugin/navigation.py [hunk 2]
```diff
@@ -275,8 +277,8 @@ def _gather_metadata(self, items: List[NavigationItem]) -> Meta:
         paths: List[str] = []
         for item in items:
             if isinstance(item, Page):
-                if Path(self.docs_dir) in Path(item.file.abs_src_path).parents:
-                    paths.append(item.file.abs_src_path)
+                if item.file.src_path in self.files.src_paths:
```
[the remark this answers is not in the document]
Comment (thejcannon) [reply mkdocs_awesome_pages_plugin/navigation.py:280]: Yeah, it would certainly need to be a hand-crafted `File`, and even then, I don't think that's something this plugin needs to bother with (garbage in, garbage out)

File: mkdocs_awesome_pages_plugin/plugin.py [hunk 3]
```diff
@@ -45,6 +47,19 @@ def __init__(self):
...
+        # Add config files to Files, so we can unconditionally load files/configs from Files
+        config_files = glob.glob(os.path.join(config["docs_dir"], f"**/{self.config['filename']}"), recursive=True)
+        for filename in config_files:
+            file = File(
+                os.path.relpath(filename, config["docs_dir"]),
+                src_dir=config["docs_dir"],
+                dest_dir=config["site_dir"],
+                use_directory_urls=config["use_directory_urls"],
+            )
+            files.append(file)
```
[the remark this answers is not in the document]
Comment (thejcannon) [reply mkdocs_awesome_pages_plugin/plugin.py:60]: I'm honestly very very new to the ecosystem, so I'm not sure I really could say. (I just like fixing the bugs I file to open source projects).

On one hand, ensuring we don't duplicate would be very nice. On the other hand, de-duplication might be expensive.

I would lean towards deduplication.

**Comment by thejcannon** [comment]
@lukasgeiter I think that should cover those comments. Happy for another look

File: mkdocs_awesome_pages_plugin/tests/e2e/test_gen_files.py [hunk 4]
```diff
@@ -48,7 +49,7 @@ def write_files(path):
 class TestGeneratedFiles(E2ETestCase):
     """See https://github.com/lukasgeiter/mkdocs-awesome-pages-plugin/issues/78."""
 
-    PLUGINS = [GeneratedFiles]
+    PLUGINS = (GeneratedFiles,)
```
Comment (lukasgeiter) [inline mkdocs_awesome_pages_plugin/tests/e2e/test_gen_files.py:52]: I'm just curious, why the change from list to tuple?

**Review by lukasgeiter — CHANGES_REQUESTED** [review]
Almost there now 😅

File: mkdocs_awesome_pages_plugin/tests/e2e/test_gen_files.py (hunk 4 above)
Comment (thejcannon) [reply mkdocs_awesome_pages_plugin/tests/e2e/test_gen_files.py:52]: So that the parent class default is immutable (same thing with function argument defaults)

File: mkdocs_awesome_pages_plugin/plugin.py [hunk 5]
```diff
@@ -50,14 +50,17 @@ def __init__(self):
...
         for filename in config_files:
-            file = File(
-                os.path.relpath(filename, config["docs_dir"]),
-                src_dir=config["docs_dir"],
-                dest_dir=config["site_dir"],
-                use_directory_urls=config["use_directory_urls"],
-            )
-            files.append(file)
+            config_path = os.path.relpath(filename, config["docs_dir"])
+            if config_path not in initial_src_paths:
```
[the remark this answers is not in the document]
Comment (lukasgeiter) [reply mkdocs_awesome_pages_plugin/plugin.py:56]: Oh yeah you're right, good catch 👍

File: mkdocs_awesome_pages_plugin/tests/e2e/test_gen_files.py (hunk 4 above)
Comment (lukasgeiter) [reply mkdocs_awesome_pages_plugin/tests/e2e/test_gen_files.py:52]: Gotcha

File: mkdocs_awesome_pages_plugin/tests/navigation/test_meta.py [hunk 6]
```diff
@@ -74,11 +74,11 @@ def populate_config_dirs(items: List[NavigationItem]):
...
         populate_config_dirs(items)
 
-        for config_dir, docs_dir in config_dirs:
+        for rel_path, abs_path in config_dirs:
             files.append(
                 File(
-                    os.path.join(config_dir, options.filename),
-                    src_dir=docs_dir,
+                    os.path.join(rel_path, options.filename),
+                    src_dir=abs_path,
```
[the remark this answers is not in the document]
Comment (thejcannon) [reply mkdocs_awesome_pages_plugin/tests/navigation/test_meta.py:81]: Looks like two wrongs make a right. Turns out the `self.page` just set the `docs_dir`/`src_dir` to `""`.

I'll push the fix

**Comment by lukasgeiter** [comment]
Released with version 2.9.2. Thanks again for your work!

**Comment by thejcannon** [comment]
Thank you for accepting it, and reviewing it
